### ISSUE TITLE
@:allow should not be inherited

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#6525]
I just hit into the following.

In heaps I have some class `h2d.Tile` with `@:allow(h2d)` so the classes in `h2d` package can access its private for implementation reason.

If I extend a `h2d` class into my own project custom class (which is not  part of the `h2d` package) then I can still access the `h2d.Tile` privates. 

I don't know if it was there from the start or if it's a regression but I think it should be fixed.

**Comment by Gama11** [comment]
I've hit this with Flixel as well - it led to some internal APIs being public by accident. I thought this was by design and transitioned to using `@:access` instead, which is "safe" in this regard.

**Comment by ncannasse** [comment]
@:allow(h2d) should be restricted to code in h2d, not to classes in other packages, We need to check superclasses for @:access (because it is inherited) but @:allow should always be checked with "current" class, not superclass

**Comment by Aurel300** [comment]
But then suppose you do a class which extends the base class with `@:allow`, but doesn't override its functions. The functions of the base class should still function, right? Even if I did:

    override public function useSomethingPrivate():Void {
      super.useSomethingPrivate();
      // stuff that doesn't use private fields
    }

It should still work. Can we tell specifically which class the function call originates from?

**Comment by Simn** [comment]
Can somebody post an actual example of what works but shouldn't work?

**Comment by Aurel300** [comment]
@Gama11 Do you agree that in your `Bar` example, `super()` and `super.anyMethod()` should still be able to access the internal methods?

By the way, imho `@:allow` and `@:access` look very hack-y. It would be nice to have proper package-level access control in Haxe.

**Comment by Simn** [comment]
I investigated a bit what's going on here and found this interesting issue: #2162

The old documentation is still available here: http://old.haxe.org/manual/acl

So the situation is that it was working as intended, but not documented accordingly. I then changed it to match the documentation and now it turns out the documentation was wrong.

Fixing this should be a matter of reverting https://github.com/HaxeFoundation/haxe/commit/539acc854dbfa5cffd9c789a2ba892a1ec53ecce

**Comment by ncannasse** [comment]
There is a problem here. Now it seems that you can't `@:allow(BaseClass)` and have sub classes access. I think that when you have a class in allow, sub classes should be given the access. But when you have a package, only classes in this package should be given the access.

**Comment by ncannasse** [comment]
See https://github.com/HeapsIO/heaps/issues/414

**Comment by ncannasse** [comment]
Regarding spec discussion:

- `@:allow(ClassName)` is specified as allowing access to a class and its subclasses, this should not be changed
- `@:allow(package)` is also well specified, it should allow access to classes defined in this package and its sub packages. But not to subclasses of these classes (unless they are also declared in this package) - hence the current issue.

Both have some kind of "inheritance" : allow on classes give access to subclasses, allow on package give access to sub packages. It's like a filter, the first one test "is the class declared in package P", the second test "is the class an instance of C"

**Comment by Simn** [comment]
What bothers me about this is that if `@:allow(some.pack.SomeModule)` allows something, I'd expect `@:allow(some.pack)` to allow at least the same because it's more general.

I get that there are two kinds of inheritance. We could make this configurable by supporting `@:allow(some.pack, { recurse: true, inherit: false })`, but then we still have to argue about what the default behavior should be. Plus it would make the access control even more complicated than it already is...

I'll think about this a bit more, there has to be a more elegant approach to all this.

**Comment by ncannasse** [comment]
> What bothers me about this is that if @:allow(some.pack.SomeModule) allows something, I'd expect @:allow(some.pack) to allow at least the same because it's more general.

Maybe, but maybe not. That's a bit second order thinking. What's important is that we are clear about what allow on both package and class does and that it actually matches usability. 

When giving access to a class, you don't have to know whether the user will extend or not the class in order to customize its behavior, so it should inherit.

When giving access to package, the intent is to keep something private for within a library usage. If it's a framework (like heaps etc) then you don't want users extending your classes to access these privates - or else they should be public or have proper acessors.

**Comment by Simn** [comment]
I acknowledge that use-case, but I'm still not sure if it should be the default/only behavior.

**Comment by ncannasse** [comment]
FYI, I confirm my previous judgment on the the issue ;)

**Comment by Simn** [comment]
If you have `@:allow(ClassName)`, should that also give you access to a sub-type `ClassName.SubClass`? I'm guessing no because it's similar to packages.

**Comment by ncannasse** [comment]
@Simn I think it should give access because we're still a subclass of this class. for packages, it will only allow if were are still in the package that was allowed.

So for instance:

@:alllow(pack) --> allow if we are in pack (or sub packages)
@:allow(Class) --> allow if we are in Class (or sub classes)

**Comment by Simn** [comment]
This issue makes me feel like an idiot because the implementation seems much harder than it should be. Let's make some tables to get an overview:

path kind | path description
--- | ---
package | all parts starts lower-case
type | last part starts upper-case, second-to-last does not
sub-type | last and second-to-last parts start upper-case
field | last part starts lower-case, second-to-last starts upper-case

Our subject paths are always of kind `field`, so we check all of these:

```
self: package.Type.field
inherited: basePackage.BaseType.field
inherited: interfacePackage.Interface.field
```

This leads to this matrix:

@:allow path kind | self | inherited
--- | --- | ---
package | packages match | disallow
type | packages + type match | packages + type match
sub-type | package + module + type match | package + module + type match
field | exact match | exact match

This means that the only special case for `inherited` is the `package` path kind, which matches your specification (I think). However, we still have to classify the paths in order to figure out if we have a `package` kind path. I think we also have to classify for correct matching of `sub-type` kind paths. I guess we should cache this classification so we don't have to determine it on every access. 

Please let me know if I'm overthinking this...

**Comment by Gama11** [comment]
>path kind | path description
>-- | --
>sub-type | last and second-to-last parts start upper-case

Hm, I think currently, `pack.SubType` is required instead of `pack.Module.SubType`. This actually tends to lead to a lot of confusion since the latter is more natural / consistent with other Haxe syntax, but fails silently.

**Comment by Simn** [comment]
Yes this has been bothering me for a while and I figured while I look at this mess anyway I might as well fix it.

**Comment by nadako** [comment]
> field	last part starts lower-case, second-to-last starts upper-case

this is not necessarily true, fields can start with upper-case though, and i don't think it's that uncommon, especially when working with c++/c# interop

**Comment by Simn** [comment]
Right, I'm kinda in denial about that possibility because it makes field and sub-type paths indistinguishable.

**Comment by ncannasse** [comment]
I think we already cover well the types/subtypes cases, so what we need to do is the following:
if allow is a package, check against the current type package without recursing on our superclass, instead of recursing as we do now.

**Comment by Simn** [comment]
Unassigned myself; issue is too hard for me.

**Comment by Simn** [comment]
Something isn't right, see Slack for details.

**Comment by RealyUniqueName** [comment]
> If I add `ignore(follow f.cf_type);` before check_constructor_access there's a test failure for 6525
> There's a `cf.cf_expr = None` check in your fix, which should always come with a `follow cf.cf_type` first to resolve `TLazy`
> The logic relies on the `if cf.cf_expr = None then false` branch being hit so that it ends up in the `|| (match c.cl_super part`
> Otherwise it raises `Not_found` which skips that

**Comment by Simn** [comment]
You didn't quote the most important part:

> Oh God that's THAT issue
